File: catalogos/rutas/puestos.py
from .rutas import catalogos
from flask import render_template, request, jsonify
from flask_login import current_user
from sqlalchemy.orm.exc import NoResultFound
from sqlalchemy import inspect, or_
import openpyxl
import collections
# import pandas as pd

from app import db
from rh.gestion_empleados.modelos.empleado import tPuesto
from catalogos.modelos.modelos import kRamo, kUA, kZonaEconomica, kTipoPlazaPuesto, kCaracterOcupacional, kTipoFuncion, kGrupo, kGrado, kNivel, kEstatusPuesto, kVigencia, kCentroTrabajo, kCentroCostos

@catalogos.route('/catalogos/puestos')
def catalogos_puestos():
    Ramo = db.session.query(kRamo).filter_by(Activo = 1).order_by(kRamo.idRamo).all()
    UA = db.session.query(kUA).filter_by(Activo = 1).order_by(kUA.idUA).all()
    ZonaEconomica = db.session.query(kZonaEconomica).filter_by(Activo = 1).order_by(kZonaEconomica.idZonaEconomica).all()
    TipoPlazaPuesto = db.session.query(kTipoPlazaPuesto).filter_by(Activo = 1).order_by(kTipoPlazaPuesto.idTipoPlazaPuesto).all()
    CaracterOcupacional = db.session.query(kCaracterOcupacional).filter_by(Activo = 1).order_by(kCaracterOcupacional.idCaracterOcupacional).all()
    TipoFuncion = db.session.query(kTipoFuncion).filter_by(Activo = 1).order_by(kTipoFuncion.idTipoFuncion).all()
    Grupo = db.session.query(kGrupo).filter_by(Activo = 1).order_by(kGrupo.idGrupo).all()
    Grado = db.session.query(kGrado).filter_by(Activo = 1).order_by(kGrado.idGrado).all()
    Nivel = db.session.query(kNivel).filter_by(Activo = 1).order_by(kNivel.idNivel).all()
    EstatusPuesto = db.session.query(kEstatusPuesto).filter_by(Activo = 1).order_by(kEstatusPuesto.idEstatusPuesto).all()
    Vigencia = db.session.query(kVigencia).filter_by(Activo = 1).order_by(kVigencia.idVigencia).all()
    CentroTrabajo = db.session.query(kCentroTrabajo).filter_by(Activo = 1).order_by(kCentroTrabajo.idCentroTrabajo).all()
    CentroCostos = db.session.query(kCentroCostos).order_by(kCentroCostos.idCentroCosto).all()

    return render_template('/puestos.html', title = 'Puestos',
                           Ramo = Ramo,
                           UA = UA,
                           ZonaEconomica = ZonaEconomica,
                           TipoPlazaPuesto = TipoPlazaPuesto,
                           CaracterOcupacional = CaracterOcupacional,
                           TipoFuncion = TipoFuncion,
                           Grupo = Grupo,
                           Grado = Grado,
                           Nivel = Nivel,
                           EstatusPuesto = EstatusPuesto,
                           Vigencia = Vigencia,
                           CentroTrabajo = CentroTrabajo,
                           CentroCostos = CentroCostos)

@catalogos.route("/catalogos/crear-puesto", methods = ["POST"])
def crear_puesto():

    mapeo_nombres = { #NombreEnFormulario : nombreEnBase
        "Ramo": "idRamo",
        "UA": "idUA",
        "ConsecutivoPuesto": "ConsecutivoPuesto",
        "CodigoPuesto": "CodigoPuesto",
        "Puesto": "Puesto",
        "ZonaEconomica": "idZonaEconomica",
        "ReferenciaTabular": "ReferenciaTabular",
        "ConsPuesto": "ConsPuesto",
        "TipoPlazaPuesto": "idTipoPlazaPuesto",
        "CaracterOcupacional": "idCaracterOcupacional",
        "TipoFuncion": "idTipoFuncion",
        "NivelSalarial": "NivelSalarial",
        "Tabulador": "Tabulador",
        "CodigoPresupuestal": "CodigoPresupuestal",
        "OrdinalCP": "OrdinalCP",
        "Grupo": "idGrupo",
        "Grado": "idGrado",
        "Nivel": "idNivel",
        "EstatusPuesto": "idEstatusPuesto",
        "Vigencia": "idVigencia",
        "FechaInicio": "FechaInicio",
        "FechaFin": "FechaFin",
        "CentroTrabajo": "idCentroTrabajo",
        "FolioSival": "FolioSival",
        "RegimenLaboral": "RegimenLaboral",
        "RemuneracionTotal": "RemuneracionTotal",
        "TitularAU": "TitularAU",
        "DeclaracionPatrimonial": "DeclaracionPatrimonial",
        "PlazasSubordinadas": "PlazasSubordinadas",
        "PuestoJefe": "PuestoJefe",
        "PresupuestalJefe": "PresupuestalJefe",
        "CentroCosto": "idCentroCosto"
    }

    puesto_data = {mapeo_nombres[key]: request.form.get(key) for key in mapeo_nombres.keys()}
    nuevo_puesto = None
    print(puesto_data)

    try:
        puesto_modificar = db.session.query(tPuesto).filter_by(ConsecutivoPuesto = puesto_data["ConsecutivoPuesto"]).one()
        puesto_modificar.update(**puesto_data)
        print("MODIFICAR")

    except NoResultFound:
        nuevo_puesto = tPuesto(**puesto_data)
        db.session.add(nuevo_puesto)
        print("AÑADIR NUEVO")

    #db.session.commit()

    return jsonify(puesto_data)

@catalogos.route("/catalogos/actualizar-busqueda-puestos", methods = ["GET"])
def actualizar_busqueda_puestos():
    texto_busqueda = request.args.get("texto_busqueda", "")
    resultados_json = []
    resultados = tPuesto.query.filter(or_(tPuesto.Puesto.ilike(f'%{texto_busqueda}%'), tPuesto.ConsPuesto.ilike(f'%{texto_busqueda}%'))).all()

    for resultado in resultados:
        if resultado is not None:
            resultado_dict = resultado.__dict__
            resultado_dict.pop("_sa_instance_state", None)  # Eliminar atributo de SQLAlchemy
            resultados_json.append(resultado_dict)

    return jsonify(resultados_json)

@catalogos.route("/catalogos/buscar-puesto", methods = ["POST"])
def buscar_puesto():
    respuesta = {}
    puesto = request.form.get("PuestoExistente")
    if puesto:
        puestos = db.session.query(tPuesto).filter(tPuesto.Puesto.contains(puesto)).all()

    else:
        puestos = []

    lista_puestos = []
    for pues in puestos:
        if pues is not None:
            puesto_dict = pues.__dict__
            puesto_dict.pop("_sa_instance_state", None)  # Eliminar atributo de SQLAlchemy
            lista_puestos.append(puesto_dict)
    
    if not lista_puestos:
        respuesta["NoEncontrado"] = True

    respuesta["ListaPuestos"] = lista_puestos

    return jsonify(respuesta)

@catalogos.route("/catalogos/cargar-archivo-puestos", methods = ["POST"])
def cargar_archivo_puestos():
    archivo = request.files.get('archivo')
    #print(archivo)

    columnas = inspect(tPuesto).all_orm_descriptors.keys()
    del columnas[32:44]

    valor_columnas =[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 32, 41, 40]
    mapeo_nombres = dict(zip(columnas, valor_columnas))

    if(archivo):
        wb = openpyxl.load_workbook(archivo)
        print(wb)
        ws = wb.active

        datos_lista = []
        datos_dict = {}
        nuevo_puesto = None

        #idEstatusPuesto, idVigencia, idCentroTrabajo

        for row_value in range(12, ws.max_row + 1):
            for key in mapeo_nombres:
                datos_dict.update({key: ws.cell(row = row_value, column = mapeo_nombres[key]).value})
                if key == "idEstatusPuesto":
                    valor = db.session.query(kEstatusPuesto).filter_by(EstatusPuesto = datos_dict["idEstatusPuesto"]).one()
                    datos_dict["idEstatusPuesto"] = valor.idEstatusPuesto
                
                if key == "idVigencia":
                    valor = db.session.query(kVigencia).filter_by(Vigencia = datos_dict["idVigencia"]).one()
                    datos_dict["idVigencia"] = valor.idVigencia

                if key == "idCentroTrabajo":
                    valor = db.session.query(kCentroTrabajo).filter_by(CentroTrabajo = datos_dict["idCentroTrabajo"]).one()
                    datos_dict["idCentroTrabajo"] = valor.idCentroTrabajo

            datos_lista.append(datos_dict.copy())

            try:
                puesto_existente = db.session.query(tPuesto).filter_by(ConsecutivoPuesto = datos_dict["ConsecutivoPuesto"]).one()
                puesto_existente.update(**datos_dict)

            except NoResultFound:
                nuevo_puesto = tPuesto(**datos_dict)
                db.session.add(nuevo_puesto)

        db.session.commit()


        return({"guardado": True})
    
    else:
        return({"guardado": False})

File: catalogos/estatico/scripts/puestos.js
$gmx(document).ready(function(){
    $("#btnCargarArchivo").click(cargar_archivo);
    $("#btnModalCrearPuesto").click(modal_crear_puesto);
    $("#btnCrearPuesto").click(crear_puesto);
    $("#btnBuscaPuesto").click(buscar_puesto);

    $("#PuestoExistente").on("focus", function(){
        $("#lista_puestos").show();
    });

    $("#PuestoExistente").on("blur", function(event){
        if(!$(event.relatedTarget).closest("#lista_puestos").length){
            $("#lista_puestos").hide();
        }
    });

    $("#PuestoExistente").on("input", function(){
        var textoBusqueda = $(this).val();
        $.ajax({
            url: "/catalogos/actualizar-busqueda-puestos",
            method: "GET",
            data: {
                texto_busqueda: textoBusqueda
            },
            success: function(response){
                actualizarListaDesplegable(response);
            }
        })
    });

    $("#FechaInicio").datepicker({changeYear: true, changeMonth: true});
    $("#FechaFin").datepicker({changeYear: true, changeMonth: true});

    configuraDatepickers("FechaInicio", "FechaFin", "");

});

function modal_crear_puesto(){
    $("#ModalCrearPuesto").modal("show");

    $("#ConsecutivoPuesto").val("");
    $("#consecutivo").val("");

    $("#Ramo").prop("disabled", false);
    $("#UA").prop("disabled", false);
    $("#CodigoPuesto").prop("readonly", false);
    $("#ZonaEconomica").prop("disabled", false);
    $("#Puesto").prop("readonly", false);
    $("#ReferenciaTabular").prop("readonly", false);
    $("#ConsPuesto").prop("readonly", false);
    $("#TipoPlazaPuesto").prop("disabled", false);
    $("#CaracterOcupacional").prop("disabled", false);
    $("#TipoFuncion").prop("disabled", false);
    $("#NivelSalarial").prop("readonly", false);
    $("#Tabulador").prop("readonly", false);
    $("#CodigoPresupuestal").prop("readonly", false);
    $("#OrdinalCP").prop("readonly", false);
    $("#Grupo").prop("disabled", false);
    $("#Grado").prop("disabled", false);
    $("#Nivel").prop("disabled", false);
    $("#EstatusPuesto").prop("disabled", false);
    $("#Vigencia").prop("disabled", false);
    $("#FechaInicio").prop("readonly", false);
    $("#FechaFin").prop("readonly", false);
    $("#CentroTrabajo").prop("disabled", false);
    $("#FolioSival").prop("readonly", false);
    $("#RegimenLaboral").prop("readonly", false);
    $("#RemuneracionTotal").prop("readonly", false);
    $("#TitularAU").prop("readonly", false);
    $("#DeclaracionPatrimonial").prop("readonly", false);
    $("#PlazasSubordinadas").prop("readonly", false);
    $("#PuestoJefe").prop("readonly", false);
    $("#PresupuestalJefe").prop("readonly", false);
    $("#CentroCosto").prop("disabled", false);

    $("#tituloModalCrearPuesto")[0].textContent = "Crear nuevo puesto";
    $("#btnCrearPuesto")[0].textContent = "Crear";

    $("#Ramo").val("0");
    $("#UA").val("0");
    $("#CodigoPuesto").val("");
    $("#ZonaEconomica").val("0");
    $("#Puesto").val("");
    $("#ReferenciaTabular").val("");
    $("#ConsPuesto").val("");
    $("#TipoPlazaPuesto").val("");
    $("#CaracterOcupacional").val("0");
    $("#TipoFuncion").val("0");
    $("#NivelSalarial").val("");
    $("#Tabulador").val("");
    $("#CodigoPresupuestal").val("");
    $("#OrdinalCP").val("");
    $("#Grupo").val("0");
    $("#Grado").val("0");
    $("#Nivel").val("0");
    $("#EstatusPuesto").val("0");
    $("#Vigencia").val("0");
    $("#FechaInicio").val("");
    $("#FechaFin").val("");
    $("#CentroTrabajo").val("0");
    $("#FolioSival").val("");
    $("#RegimenLaboral").val("");
    $("#RemuneracionTotal").val("");
    $("#TitularAU").val("");
    $("#DeclaracionPatrimonial").val("");
    $("#PlazasSubordinadas").val("");
    $("#PuestoJefe").val("");
    $("#PresupuestalJefe").val("");
    $("#CentroCosto").val("0");
}

function actualizarListaDesplegable(resultados){
    var listaDesplegable = $("#lista_puestos");
    listaDesplegable.empty(); // Vaciar la lista desplegable antes de agregar nuevos elementos
    resultados.forEach(function(resultado){
        var nuevoElemento = $('<div>', {
            'class': 'dropdown-item form-group',
            'html': $('<a>', {
                'href': 'javascript:void(0)', // El href está configurado para evitar que la página se recargue
                'data-id': resultado.ConsPuesto,
                'data-texto': resultado.Puesto,
                'text': resultado.Puesto,
                'click': function(){
                    var textoSeleccionado = $(this).data('texto');
                    $("#PuestoExistente").val(textoSeleccionado);
                    $("#lista_puestos").hide();
                }
            })
        });
        listaDesplegable.append(nuevoElemento);
    });
}

function crear_puesto(){
    if(validarFormulario($("#formularioCrearPuesto")).valido){
        $.ajax({
            async: false,
            type: "POST",
            url: "/catalogos/crear-puesto",
            data: $("#formularioCrearPuesto, #consecutivo").serialize(),
            success: function(data){
                if(data){
                    abrirModal("Información guardada", "Operación realizada con éxito.", "");
                    $("#ModalCrearPuesto").modal("hide");
                }
            }
        });
    }
}

function buscar_puesto(){
    $.ajax({
        async: false,
        type: "POST",
        url: "/catalogos/buscar-puesto",
        data: $("#formularioBuscarPuesto").serialize(),
        success: function(respuesta){
            if(respuesta.NoEncontrado){
                abrirModal("No encontrado", "No se encontraron coincidencias.", "");
            }else{
                $("#tablaResultadosPuestos").show();
                $("#tablaResultadosPuestos tbody").empty();
                var cont = 1;

                respuesta.ListaPuestos.forEach(function(puesto){
                    text = `
                        <tr>
                            <input type="hidden" id="Ramo${cont}" value="${puesto.idRamo}"></input>
                            <input type="hidden" id="UA${cont}" value="${puesto.idUA}"></input>
                            <input type="hidden" id="ConsecutivoPuesto${cont}" value="${puesto.ConsecutivoPuesto}"></input>
                            <input type="hidden" id="CodigoPuesto${cont}" value="${puesto.CodigoPuesto}"></input>
                            <input type="hidden" id="ZonaEconomica${cont}" value="${puesto.idZonaEconomica}"></input>
                            <input type="hidden" id="ReferenciaTabular${cont}" value="${puesto.ReferenciaTabular}"></input>
                            <input type="hidden" id="ConsPuesto${cont}" value="${puesto.ConsPuesto}"></input>
                            <input type="hidden" id="TipoPlazaPuesto${cont}" value="${puesto.idTipoPlazaPuesto}"></input>
                            <input type="hidden" id="CaracterOcupacional${cont}" value="${puesto.idCaracterOcupacional}"></input>
                            <input type="hidden" id="TipoFuncion${cont}" value="${puesto.idTipoFuncion}"></input>
                            <input type="hidden" id="NivelSalarial${cont}" value="${puesto.NivelSalarial}"></input>
                            <input type="hidden" id="Tabulador${cont}" value="${puesto.Tabulador}"></input>
                            <input type="hidden" id="CodigoPresupuestal${cont}" value="${puesto.CodigoPresupuestal}"></input>
                            <input type="hidden" id="OrdinalCP${cont}" value="${puesto.OrdinalCP}"></input>
                            <input type="hidden" id="Grupo${cont}" value="${puesto.idGrupo}"></input>
                            <input type="hidden" id="Grado${cont}" value="${puesto.idGrado}"></input>
                            <input type="hidden" id="EstatusPuesto${cont}" value="${puesto.idEstatusPuesto}"></input>
                            <input type="hidden" id="Vigencia${cont}" value="${puesto.idVigencia}"></input>
                            <input type="hidden" id="FechaInicio${cont}" value="${puesto.FechaInicio}"></input>
                            <input type="hidden" id="FechaFin${cont}" value="${puesto.FechaFin}"></input>
                            <input type="hidden" id="CentroTrabajo${cont}" value="${puesto.idCentroTrabajo}"></input>
                            <input type="hidden" id="FolioSival${cont}" value="${puesto.FolioSival}"></input>
                            <input type="hidden" id="RegimenLaboral${cont}" value="${puesto.RegimenLaboral}"></input>
                            <input type="hidden" id="RemuneracionTotal${cont}" value="${puesto.RemuneracionTotal}"></input>
                            <input type="hidden" id="TitularAU${cont}" value="${puesto.TitularAU}"></input>
                            <input type="hidden" id="DeclaracionPatrimonial${cont}" value="${puesto.DeclaracionPatrimonial}"></input>
                            <input type="hidden" id="PlazasSubordinadas${cont}" value="${puesto.PlazasSubordinadas}"></input>
                            <input type="hidden" id="PuestoJefe${cont}" value="${puesto.PuestoJefe}"></input>
                            <input type="hidden" id="PresupuestalJefe${cont}" value="${puesto.PresupuestalJefe}"></input>
                            <input type="hidden" id="CentroCosto${cont}" value="${puesto.idCentroCosto}"></input>
                            <td>
                                <input type="text" id="Puesto${cont}" value="${puesto.Puesto}" class="form-control" readonly></input>
                            </td>

                            <td>
                                <div style="display: block;">
                                    <button type="button" class="btn btn-primary btn-sm" id="Editar_Aceptar${cont}" onclick="editar_puesto(${cont})"><span class="glyphicon glyphicon-pencil" aria-hidden="true"></span></button>
                                </div>
                            </td>
                        </tr>
                    `;
                    $("#tablaResultadosPuestos tbody").append(text);
                    cont++;
                });
            }
        }
    });
}

function editar_puesto(consecutivo){
    $("#ModalCrearPuesto").modal("show");

    $("#consecutivo").val(consecutivo);

    $("#Ramo").prop("disabled", false);
    $("#UA").prop("disabled", false);
    $("#CodigoPuesto").prop("readonly", false);
    $("#ZonaEconomica").prop("disabled", false);
    $("#Puesto").prop("readonly", false);
    $("#ReferenciaTabular").prop("readonly", false);
    $("#ConsPuesto").prop("readonly", false);
    $("#TipoPlazaPuesto").prop("disabled", false);
    $("#CaracterOcupacional").prop("disabled", false);
    $("#TipoFuncion").prop("disabled", false);
    $("#NivelSalarial").prop("readonly", false);
    $("#Tabulador").prop("readonly", false);
    $("#CodigoPresupuestal").prop("readonly", false);
    $("#OrdinalCP").prop("readonly", false);
    $("#Grupo").prop("disabled", false);
    $("#Grado").prop("disabled", false);
    $("#Nivel").prop("disabled", false);
    $("#EstatusPuesto").prop("disabled", false);
    $("#Vigencia").prop("disabled", false);
    $("#FechaInicio").prop("readonly", false);
    $("#FechaFin").prop("readonly", false);
    $("#CentroTrabajo").prop("disabled", false);
    $("#FolioSival").prop("readonly", false);
    $("#RegimenLaboral").prop("readonly", false);
    $("#RemuneracionTotal").prop("readonly", false);
    $("#TitularAU").prop("readonly", false);
    $("#DeclaracionPatrimonial").prop("readonly", false);
    $("#PlazasSubordinadas").prop("readonly", false);
    $("#PuestoJefe").prop("readonly", false);
    $("#PresupuestalJefe").prop("readonly", false);
    $("#CentroCosto").prop("disabled", false);

    $("#tituloModalCrearPuesto")[0].textContent = "Editar puesto";
    $("#btnCrearPuesto")[0].textContent = "Editar";

    FechaInicio = convertirFechaParaVisualizacion($("#FechaInicio" + consecutivo).val());
    FechaFin = convertirFechaParaVisualizacion($("#FechaFin" + consecutivo).val())

    $("#Ramo").val($("#Ramo" + consecutivo).val());
    $("#UA").val($("#UA" + consecutivo).val());
    $("#ConsecutivoPuesto").val($("#ConsecutivoPuesto" + consecutivo).val())
    $("#CodigoPuesto").val($("#CodigoPuesto" + consecutivo).val());
    $("#ZonaEconomica").val($("#ZonaEconomica" + consecutivo).val());
    $("#Puesto").val($("#Puesto" + consecutivo).val());
    $("#ReferenciaTabular").val($("#ReferenciaTabular" + consecutivo).val());
    $("#ConsPuesto").val($("#ConsPuesto" + consecutivo).val());
    $("#TipoPlazaPuesto").val($("#TipoPlazaPuesto" + consecutivo).val());
    $("#CaracterOcupacional").val($("#CaracterOcupacional" + consecutivo).val());
    $("#TipoFuncion").val($("#TipoFuncion" + consecutivo).val());
    $("#NivelSalarial").val($("#NivelSalarial" + consecutivo).val());
    $("#Tabulador").val($("#Tabulador" + consecutivo).val());
    $("#CodigoPresupuestal").val($("#CodigoPresupuestal" + consecutivo).val());
    $("#OrdinalCP").val($("#OrdinalCP" + consecutivo).val());
    $("#Grupo").val($("#Grupo" + consecutivo).val());
    $("#Grado").val($("#Grado" + consecutivo).val());
    $("#Nivel").val($("#Nivel" + consecutivo).val());
    $("#EstatusPuesto").val($("#EstatusPuesto" + consecutivo).val());
    $("#Vigencia").val($("#Vigencia" + consecutivo).val());
    $("#FechaInicio").val(FechaInicio);
    $("#FechaFin").val(FechaFin);
    $("#CentroTrabajo").val($("#CentroTrabajo" + consecutivo).val());
    $("#FolioSival").val($("#FolioSival" + consecutivo).val());
    $("#RegimenLaboral").val($("#RegimenLaboral" + consecutivo).val());
    $("#RemuneracionTotal").val($("#RemuneracionTotal" + consecutivo).val());
    $("#TitularAU").val($("#TitularAU" + consecutivo).val());
    $("#DeclaracionPatrimonial").val($("#DeclaracionPatrimonial" + consecutivo).val());
    $("#PlazasSubordinadas").val($("#PlazasSubordinadas" + consecutivo).val());
    $("#PuestoJefe").val($("#PuestoJefe" + consecutivo).val());
    $("#PresupuestalJefe").val($("#PresupuestalJefe" + consecutivo).val());
    $("#CentroCosto").val($("#CentroCosto" + consecutivo).val());

}

function cargar_archivo(){
    const archivo = new FormData();
    archivo.append('archivo', $("#Archivo")[0].files[0]);

    $.ajax({
        async: false,
        type: "POST",
        url: "/catalogos/cargar-archivo-puestos",
        data: archivo,
        enctype: "multipart/form-data",
        contentType: false,
        processData: false,
        success: function(data){
            if(data.guardado){
                abrirModal("Datos cargados", "Los datos se han cargado correctamente.", "recargar");
            }else{
                abrirModal("Datos no cargados", "No se seleccionó un archivo.", "");
            }
        }
    });
}

function configuraDatepickers(idInicio, idFin, data) {
    $(`#${idInicio}${data}, #${idFin}${data}`).datepicker({
        // $(`#fechaInicioFormateada${data}, #fechaFinFormateada${data}`).datepicker({
        changeYear: true,
        changeMonth: true,
        beforeShow: function (input, inst) {
            var fechaLimite = this.id === `${idInicio}${data}` ? $(`#${idFin}${data}`).datepicker("getDate") : $(`#${idInicio}${data}`).datepicker("getDate");
            if (this.id === `${idInicio}${data}`) {
                if (fechaLimite) {
                    $(this).datepicker("option", "maxDate", fechaLimite);
                } else {
                    //$(this).datepicker("option", "maxDate", null);
                }
            } else {
                if (fechaLimite) {
                    $(this).datepicker("option", "minDate", fechaLimite);
                }
                else {
                    //$(this).datepicker("option", "minDate", null);
                }
            }
        },
        onSelect: function (dateText) {
            var fechaInicial = $(`#${idInicio}${data}`).datepicker("getDate");
            var fechaFinal = $(`#${idFin}${data}`).datepicker("getDate");

            if (fechaFinal && fechaInicial) {
                if (fechaInicial > fechaFinal) {
                    abrirModal("Error", "La fecha de inicio no puede ser mayor que la fecha de fin.", "");
                    $(this).val("");

                } else {
                    // var fechaInicioISO = $.datepicker.formatDate('yy-mm-dd', fechaInicial);
                    // var fechaFinISO = $.datepicker.formatDate('yy-mm-dd', fechaFinal);

                    // $.ajax({
                    //     async: false,
                    //     type: "POST",
                    //     url: "/RH/obtener_quincenas_entre_fechas",
                    //     data: {
                    //         "FechaInicio": fechaInicioISO,
                    //         "FechaFin": fechaFinISO
                    //     },
                    //     success: function (data) {
                    //         $("#NumeroQuincena").val(data)
                    //      }
                    // });
                }
            }
        }
    });
}